contracts/Rewards.sol

```diff
@@ -0,0 +1,212 @@
+      totalIneligibleWeight = iWeight;
+    }
+    // In any case, update their accumulator and weight
+    o.accumulated = acc;
+    o.weight = newWeight;
+    operatorRewards[operator] = o;
+  }
+
+  /// @notice Set the amount of withdrawable tokens to zero
+  /// and return the previous withdrawable amount.
+  /// @dev Does not update the withdrawable amount,
+  /// but should usually be accompanied by an update.
+  function withdrawOperatorRewards(address operator)
+    internal
+    returns (uint96 withdrawable)
+  {
+    OperatorRewards storage o = operatorRewards[operator];
+    withdrawable = o.available;
+    o.available = 0;
+  }
+
+  /// @notice Set the given operators as ineligible for rewards.
+  /// The operators can restore their eligibility at the given time.
+  function setIneligible(address[] memory operators, uint256 until) internal {
+    OperatorRewards memory o = OperatorRewards(0, 0, 0, 0);
+    uint96 globalAcc = globalRewardAccumulator;
+    uint96 accrued = 0;
+    uint256 ineligibleWeightSum = 0;
+    // Record ineligibility as seconds after contract creation
+    uint32 _until = uint32(until - ineligibleOffsetStart);
+
+    for (uint256 i = 0; i < operators.length; i++) {
+      address operator = operators[i];
+      OperatorRewards storage r = operatorRewards[operator];
+      o.available = r.available;
+      o.accumulated = r.accumulated;
+      o.ineligibleUntil = r.ineligibleUntil;
+      o.weight = r.weight;
+
+      if (o.ineligibleUntil != 0) {
+        // If operator is already ineligible,
+        // don't earn rewards or shorten its ineligibility
+        if (o.ineligibleUntil < _until) {
+          o.ineligibleUntil = _until;
+        }
+      } else {
+        // The operator becomes ineligible -> earn rewards
+        o.ineligibleUntil = _until;
+        accrued = (globalAcc - o.accumulated) * uint96(o.weight);
+        o.available += accrued;
+        ineligibleWeightSum += uint256(o.weight);
+      }
+      o.accumulated = globalAcc;
+
+      r.available = o.available;
+      r.accumulated = o.accumulated;
+      r.ineligibleUntil = o.ineligibleUntil;
+      r.weight = o.weight;
+    }
+    totalIneligibleWeight += uint32(ineligibleWeightSum);
+  }
+
+  /// @notice Restore the given operator's eligibility for rewards.
+  function restoreEligibility(address operator) internal {
+    OperatorRewards memory o = operatorRewards[operator];
+    // solhint-disable-next-line not-rely-on-time
+    require(canRestoreRewardEligibility(operator), "Operator still ineligible");
+    uint96 acc = globalRewardAccumulator;
+    o.accumulated = acc;
+    o.ineligibleUntil = 0;
+    totalIneligibleWeight -= o.weight;
+    operatorRewards[operator] = o;
+  }
+
+  /// @notice Calculate the amount of tokens the operator could withdraw now
+  /// if its current weight is `oldWeight`.
+  /// Does not update state, but returns what the results would be.
+  function availableRewards(address operator, uint256 oldWeight)
```

Comment:
Should we delete this function? It is not exposed in `SortitionPool` and we do not have tests for it. Every operator can do the math on their own after all. It also does not take into account the ineligible state as far as I understand.

Reply:
Deleted 👍